File: baseStation/src/commands.cpp
#include "commands.h"
#include "main.h"
#include "eprom.h"
#include "radio.h"

#include <iostream>
#include <queue>
#include <fstream>
#include <sstream>

using std::cout;
using std::endl;

TYPE_WIND_UPDATE commands::eeprom::wind::wind_update = 250;
TYPE_WIND_STORAGE commands::eeprom::wind::wind_storage = 2000;

TYPE_AVG_WIND_UPDATE commands::eeprom::wind::avg_wind_update = 1000;
TYPE_AVG_WIND_STORAGE commands::eeprom::wind::avg_wind_storage = 120000;

TYPE_ALTITUDE commands::eeprom::pressure::altitude = 0;

TYPE_REFRESH_TIME commands::eeprom::refresh_time = 30000;
TYPE_LISTEN_TIME commands::eeprom::listen_time = 500;


bool commands::status::just_started;
bool commands::status::charging;
float commands::status::battery;
float commands::status::battery_temp;
uint8_t commands::status::lost_packets;


uint16_t commands::weather::rawWindSpeed;
uint8_t  commands::weather::rawWindDirection;
uint16_t commands::weather::rawMaxWindSpeed;
uint8_t  commands::weather::rawMaxWindDirection;
uint16_t commands::weather::rawAverageWindSpeed;
uint16_t commands::weather::rawAverageWindDirection;

uint8_t commands::weather::rawHumidity;
int16_t commands::weather::rawTemperature;
uint8_t commands::weather::rawRainFall;
uint16_t commands::weather::rawPressure;

//==================== Command Queue ====================

struct Command {
    //pointer to command array
    uint8_t* command;
    //size of command array
    uint8_t size;
    //expects a reply for the command
    bool expectReply;
};

std::queue<Command> qCommands;
commands::EEPROM_Got eepromSuccess;
commands::Status_got statusSuccess;

Command addCommand(uint8_t size, bool expectReply) {
    Command c;
    c.expectReply = expectReply;
    c.size = size;
    c.command = new uint8_t[size];
    qCommands.push(c);
    return c;
}

void deleteCommand(Command c) {
    if(c.command) {
        delete[] c.command;
        c.command = NULL;
    }
    return;
}

//==================== Command functions ====================

template<typename T>
void commands::setEEPROM(EEPROM_Variable variable, T value) {
    Command c = addCommand(2 + sizeof(T), false);

    c.command[0] = COMMAND_SET_VALUE;
    c.command[1] = variable;
    global::set<T>(c.command + 2, value);
}
template void commands::setEEPROM<uint8_t>(EEPROM_Variable variable, uint8_t value);
template void commands::setEEPROM<uint16_t>(EEPROM_Variable variable, uint16_t value);
template void commands::setEEPROM<uint32_t>(EEPROM_Variable variable, uint32_t value);

void commands::getEEPROM(commands::EEPROM_Got success) {
    Command c = addCommand(1, true);
    c.command[0] = COMMAND_GET_SETTINGS;
    eepromSuccess = success;
}

void commands::getStatus(commands::Status_got success) {
    Command c = addCommand(1, true);
    *c.command = COMMAND_GET_STATUS;
    statusSuccess = success;
}

void commands::resetWeather() {
    Command c = addCommand(1, false);
    *c.command = COMMAND_RESET;
}

void commands::resetStatus() {
    Command c = addCommand(1, false);
    *c.command = COMMAND_RESET_STATUS;
}

void commands::restart() {
    Command c = addCommand(1, false);
    *c.command = COMMAND_RESTART;
}

void commands::resetEEPROM() {
    Command c = addCommand(1, false);
    *c.command = COMMAND_RESET_SETTINGS;
}


//==================== Process COmmands ====================

#define WIND_TICK_SIZE 1
#define WIND_READ_TIME_SIZE 2

#define WIND_AVG_UPDATE_TIME_SIZE 2
#define WIND_AVG_STORAGE_TIME_SIZE 4

#define PRES_ALTITUDE_SIZE 2

#define REFRESH_TIME_SIZE 4

#define RESETS_SIZE 1

/**
 * Get status from commands::getStatus()
 * 
 * @param data location of data
 * 
 */
void readStatus(const uint8_t* data) {
    const uint8_t BOOLEANS = 0;

    const uint8_t JUST_STARTED_BIN = 0;
    const uint8_t CHARGING_BIN = 1;

    const uint8_t BATTERY = 1;
    const uint8_t BATT_TEMP = 2;
    const uint8_t LOST_PACKETS = 4;

    commands::status::just_started = global::getBool(data + BOOLEANS, JUST_STARTED_BIN);
    commands::status::charging = global::getBool(data + BOOLEANS, CHARGING_BIN);

    // load and convert the battery voltage/temperature
    uint8_t batt = global::get<uint8_t>(data + BATTERY);
    commands::status::battery = (batt / 100.0) + 1.65;
    int16_t batt_temp = global::get<int16_t>(data + BATT_TEMP);
    commands::status::battery_temp = batt_temp / 10.0;

    commands::status::lost_packets = global::get<uint8_t>(data + LOST_PACKETS);

    if(statusSuccess) {
        (*statusSuccess)();
    }
}

/**
 * Get EEPROM from commands::getEEPROM()
 */
bool readEEPROM(const uint8_t* data) {
    const uint8_t WEATHER_INTERVAL = 0;
    const uint8_t LISTEN_TIME = 4;
    const uint8_t ALTITUDE = 6;
    const uint8_t AVG_WIND_UPDATE = 8;
    const uint8_t AVG_WIND_STORAGE = 10;
    const uint8_t WIND_UPDATE = 14;
    const uint8_t WIND_STORAGE = 16;

    commands::eeprom::refresh_time = global::get<TYPE_REFRESH_TIME>(data + WEATHER_INTERVAL);
    commands::eeprom::listen_time = global::get<TYPE_LISTEN_TIME>(data + LISTEN_TIME);
    commands::eeprom::pressure::altitude = global::get<TYPE_ALTITUDE>(data + ALTITUDE);
    commands::eeprom::wind::avg_wind_update = global::get<TYPE_AVG_WIND_UPDATE>(data + AVG_WIND_UPDATE);
    commands::eeprom::wind::avg_wind_storage = global::get<TYPE_AVG_WIND_STORAGE>(data + AVG_WIND_STORAGE);
    commands::eeprom::wind::wind_update = global::get<TYPE_WIND_UPDATE>(data + WIND_UPDATE);
    commands::eeprom::wind::wind_storage = global::get<TYPE_WIND_STORAGE>(data + WIND_STORAGE);

    if(eepromSuccess) {
        (*eepromSuccess)();
    }
}

bool commands::loadCommand(uint8_t* packet, uint8_t size) {
    uint8_t i = 0;
    bool expectResponse = false;

    if(!qCommands.empty()) {
        Command com = qCommands.front();
        uint8_t ind = size < com.size ? size : com.size;
        while(ind--) {
            packet[ind] = com.command[ind];
        }
        if(com.expectReply) {
            expectResponse = true;
        }
        deleteCommand(com);
        qCommands.pop();
    }

    return expectResponse;
}


void loadRawWeather(const uint8_t* data) {

    commands::weather::rawWindDirection = global::get<uint8_t>(data + 0);
    commands::weather::rawMaxWindDirection = global::get<uint8_t>(data + 1);
    commands::weather::rawAverageWindDirection = global::get<uint16_t>(data + 2);
    commands::weather::rawWindSpeed = global::get<uint16_t>(data + 4);
    commands::weather::rawMaxWindSpeed = global::get<uint16_t>(data + 6);
    commands::weather::rawAverageWindSpeed = global::get<uint16_t>(data + 8);

    commands::weather::rawRainFall = global::get<uint8_t>(data + 10);
    commands::weather::rawTemperature = global::get<uint16_t>(data + 11);
    commands::weather::rawHumidity = global::get<uint8_t>(data + 13);
    commands::weather::rawPressure = global::get<uint16_t>(data + 14);
}


void commands::getReply(const uint8_t* packet, uint8_t size) {
    uint8_t i = 0;

    uint8_t command = packet[0];

    if(command == COMMAND_GET_WEATHER) {
        loadRawWeather(packet + 1);
    } else if(command == COMMAND_GET_SETTINGS) {
        //TODO: Process get settings Command
    } else if(command == COMMAND_GET_STATUS) {
        readStatus(packet + 1);
    }
}

File: basestationplugin/radio/commands.py
import queue
import enum

import threading

from radio.packet import Packet, Command


class Callable_Command(threading.Thread):

    def __init__(self, command, args=None, kwargs=None):
        super().__init__()
        self._command = command
        self._args = args
        self._kwargs = kwargs

    def run(self):
        if callable(self._command):
            self._command(*self._args, **self._kwargs)


class EEPROM_Variable(enum.Enum):
    REFRESH_TIME = 0xA1
    LISTEN_TIME = 0x17
    ALTITUDE = 0xC0
    AVG_WIND_UPDATE = 0xD0
    AVG_WIND_STORAGE = 0xD3
    WIND_UPDATE = 0xDA
    WIND_STORAGE = 0xDD


class Command_Bytes(enum.Enum):
    GET_WEATHER = 0xA0
    GET_SETTINGS = 0xA5
    GET_STATUS = 0xAB

    RESET = 0xB0
    RESET_STATUS = 0xB3

    RESET_SETTINGS = 0xC0
    RESTART = 0xCA

    SET_SETTINGS = 0xD5


class RawWeather:

    def __init__(self, *args, **kwargs):
        self.windDirection = kwargs.get('windDirection', 0)
        self.maxWindDirection = kwargs.get('maxWindDirection', 0)
        self.averageWindDirection = kwargs.get('averageWindDirection', 0)

        self.windSpeed = kwargs.get('windSpeed', 0)
        self.maxWindSpeed = kwargs.get('maxWindSpeed', 0)
        self.averageWindSpeed = kwargs.get('averageWindSpeed', 0)

        self.rainFall = kwargs.get('rainFall', 0)
        self.temperature = kwargs.get('temperature', 0)
        self.humidity = kwargs.get('humidity', 0)
        self.pressure = kwargs.get('pressure', 0)


class Commands:

    def __init__(self):
        self._commands = queue.Queue()
        self._eeprom_receive = queue.Queue()
        self._status_receive = queue.Queue()
        self._rawWeather = RawWeather()

    @property
    def raw_weather(self):
        return self._rawWeather

    def addCommand(self, command: Command):
        self._commands.put(command)

    def popCommand(self):
        return self._commands.get()

    def moreCommands(self):
        return not self._commands.empty()

    def setEEPROM(self, variable: int, value):
        packet = Packet(size=6)

        packet.set8(0, Command_Bytes.SET_SETTINGS)
        packet.set8(1, variable)
        packet.set32(2, value)

        self.addCommand(Command(packet, False))

    def getEEPROM(self, on_received: callable):
        packet = Packet(packet=[Command_Bytes.GET_SETTINGS])
        self.addCommand(Command(packet, True))
        self._eeprom_receive.put(on_received)

    def getStatus(self, on_received: callable):
        packet = Packet(packet=[Command_Bytes.GET_STATUS])
        self.addCommand(Command(packet, True))
        self._status_receive.put(on_received)

    def resetWeather(self):
        self.addCommand(Command((Packet(packet=[Command_Bytes.RESET]))))

    def resetStatus(self):
        self.addCommand(Command((Packet(packet=[Command_Bytes.RESET_STATUS]))))

    def restart(self):
        self.addCommand(Command((Packet(packet=[Command_Bytes.RESTART]))))

    def _loadRawWeather(self, packet: Packet):
        self._rawWeather.windDirection = packet.get8(1)
        self._rawWeather.maxWindDirection = packet.get8(2)
        self._rawWeather.averageWindDirection = packet.get16(3)
        self._rawWeather.windSpeed = packet.get16(5)
        self._rawWeather.maxWindSpeed = packet.get16(7)
        self._rawWeather.averageWindSpeed = packet.get16(9)

        self._rawWeather.rainFall = packet.get8(11)
        self._rawWeather.temperature = packet.get16(12)
        self._rawWeather.humidity = packet.get8(14)
        self._rawWeather.pressure = packet.get16(15)

    def _readEEPROM(self, packet: Packet):
        eeprom = dict(
            refresh_time=packet.get32(1),
            listen_time=packet.get16(5),
            altitude=packet.get16(7),
            avg_wind_update=packet.get16(9),
            avg_wind_storage=packet.get32(11),
            wind_update=packet.get16(15),
            wind_storage=packet.get16(17)
        )

        on_eeprom = self._eeprom_receive.get()
        if callable(on_eeprom):
            thread = Callable_Command(on_eeprom, [eeprom])
            thread.start()

    def _read_status(self, packet: Packet):
        status = dict(
            just_started=packet.getBool(1, 0),
            charging=packet.getBool(1, 1),
            battery=(packet.get8(2) / 100.0) + 1.65,
            battery_temp=packet.get16(3) / 10.0,
            lost_packets=packet.get8(5)
        )

        on_status = self._status_receive.get()
        if callable(on_status):
            thread = Callable_Command(on_status, [status])
            thread.start()

    def getReply(self, packet: Packet):
        command = packet.get8(0)

        if command == Command_Bytes.GET_WEATHER:
            self._loadRawWeather(packet)
        elif command == Command_Bytes.GET_SETTINGS:
            self._readEEPROM(packet)
        elif command == Command_Bytes.GET_STATUS:
            self._read_status(packet)


commands = Commands()
